torch/lib/c10d/reducer.cpp

```diff
@@ -422,7 +422,7 @@ void Reducer::mark_variable_ready_dense(VariableIndex index) {
       // bucket_view. If gradient_as_bucket_view_ is set as true, let grad point
       // to bucket_view. If grad has already been set as views of buckets in
       // previous iterations, no copy is needed.
-      if (!grad.is_alias_of(bucket_view)) {
+      if (!c10d::viewSameMemory(grad, bucket_view)) {
```

Comment:
@albanD if I add a log before this line 
```
      if (process_group_->getRank() == 0 && num_iterations_ == 2) {
        for (size_t i = 0; i < replica.bucket_views_in.size(); i++) {
          std::cout << bucket_index.intra_bucket_index << "th grad is alias of "
            << i << "th bucket view, " << grad.is_alias_of(replica.bucket_views_in.at(i))
            << ", they view the exact same memory: "
            << c10d::viewSameMemory(grad, replica.bucket_views_in.at(i))
            << "\n";
        }
      }
```
and run a unit test with grad_as_bucket_view = true,  it will print out logs like this 
```
0th grad is alias of 0th bucket view, 1, they view the exact same memory: 1
0th grad is alias of 1th bucket view, 1, they view the exact same memory: 0
0th grad is alias of 2th bucket view, 1, they view the exact same memory: 0
1th grad is alias of 0th bucket view, 1, they view the exact same memory: 0
1th grad is alias of 1th bucket view, 1, they view the exact same memory: 1
1th grad is alias of 2th bucket view, 1, they view the exact same memory: 0
2th grad is alias of 0th bucket view, 1, they view the exact same memory: 0
2th grad is alias of 1th bucket view, 1, they view the exact same memory: 0
2th grad is alias of 2th bucket view, 1, they view the exact same memory: 1
```
I’m expecting 0th grad will be alias of 0th bucket_view only. But the log shows the 0th grad is alias of all bucket_views in the bucket buffer tensor. 

Although it is not resulting in numeric error for current case we've known. But we feel like is_alias_of is not what we want originally. So trying to write a helper to check whether the two tensors are actually viewing the exact same memory.

Reply:
And what should it return for partial overlap? True or False?

Reply:
we want it to return False if two tensors are partially overlapped. Basically we want two tensors to return True only when they are exactly the same. (Will tensor1==tensor2 work...?)

Reply:
In that case, yes you can check for equality. As Tensor in c++ are shared pointers and a single object in python.

Note though that `t1 == t1.view_as(t1)` will be false because the result of the view op is a new Tensor. That is what you want right?